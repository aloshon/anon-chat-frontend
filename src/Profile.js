import React, { useContext, useEffect } from "react";
import UserContext from "./UserContext";
import Container from "react-bootstrap/Container";
import Button from "react-bootstrap/Button";
import {useSelector, useDispatch} from "react-redux";
import BlockedUser from "./BlockedUser";
import AddUserToBlockListForm from "./Forms/AddUserToBlockListForm";
import {toggleDarkMode} from "./Actions/actionCreators";
import AnonChatApi from "./api";
import {useHistory} from "react-router-dom";
import {getBlockList} from './Actions/actionCreators';

/**
 * Profile renders component that displays user's username
 * and list of blocked users
 * User can toggle between dark and light mode
 * and delete their accounts here.
 */
const Profile = () => {
    const {user, logout} = useContext(UserContext);
    const dispatch = useDispatch();
    const history = useHistory();
    const blockList = useSelector(state => state.blockList);
    const darkMode = useSelector(state => state.darkMode);
   
    // Get the user's blockList on render after user is defined
    useEffect(() =>{
        if(user){
            getBlockList(dispatch, user.blockList);
        }
    }, [user, dispatch]);

    // scroll user to top of page
    useEffect(() => {
        window.scrollTo(0, 0)
    }, []);

    const handleDelete = () => {
        if(window.confirm("Delete your profile?")){
            logout();
            AnonChatApi.deleteUser();
            history.push('/signup')
        } else{
            return
        }
    }

    // If user is not logged in or loaded yet, return loading
    while(!user) return <h1>Loading...</h1>;

    return(
        <Container>
            <h1>Username: {user.username}</h1><br/>
            <div className='custom-control custom-switch'>
                <input
                type='checkbox'
                className='custom-control-input'
                id='customSwitchesChecked'
                onChange={() => toggleDarkMode(dispatch)}
                checked={darkMode.isDarkMode ? true : false}
                />
                <label className='custom-control-label' htmlFor='customSwitchesChecked'>
                {darkMode.isDarkMode ? "dark mode" : "light mode"}
                </label>
            </div>
            <AddUserToBlockListForm blockList={blockList} />
            <h5 style={{marginTop: "90px"}}>Block List</h5>
            <Container>
                {blockList.length !== 0 ? blockList.map(b => (
                    <BlockedUser 
                    key={b.id}
                    blockedUser={b.blocked_username}
                    />
                )) : <h2>No users blocked</h2>}
            </Container>
            <Button
            style={{ float: "left", marginTop: "30px" }}
            variant="danger" 
            onClick={handleDelete}>
                Delete Profile
            </Button>
        </Container>
    )
}

export default Profile;