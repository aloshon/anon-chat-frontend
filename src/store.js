import  { composeWithDevTools } from "redux-devtools-extension";
import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import autoMergeLevel2 from "redux-persist/lib/stateReconciler/autoMergeLevel2";

import thunk from "redux-thunk";
import rootReducer from "./Reducers/rootReducer";
import { createStore, applyMiddleware } from "redux";

// only keep the darkMode reducer persistent in local storage
const persistConfig = {
  key: "root",
  whitelist:['darkMode'],
  storage,
  stateReconciler: autoMergeLevel2
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = createStore(
  persistedReducer,
  composeWithDevTools(
    applyMiddleware(thunk),
  )
);

export const persistedStore = persistStore(store);