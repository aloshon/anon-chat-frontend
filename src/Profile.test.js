import { render, screen } from '@testing-library/react';
import Profile from './Profile';
import { Provider } from "react-redux";
import { store } from "./store"
import {BrowserRouter} from 'react-router-dom';
import UserContext from "./UserContext";


describe("Profile should render username and block list", () => {
    const user = {
        id: 1,
        username: "user1",
        isAdmin: false,
        blockList: [{
            id: 1,
            blocked_username: "test"
        }]
    }

    function logout(user){
        user = undefined;
        return user
    }

    const {asFragment} = render(
      <Provider store={store}>
        <BrowserRouter>
            <UserContext.Provider value={{user, logout}}>
                <Profile />
            </UserContext.Provider>
        </BrowserRouter>
      </Provider>
    );

    test("renders username and blocked_username in block list", () => {
        const username = screen.getByText("Username: user1");
        const blockedUser = screen.getByText("test");
        expect(username).toBeInTheDocument();
        expect(blockedUser).toBeInTheDocument();
    });
  
    test("matches snapshot", () => {
      expect(asFragment()).toMatchSnapshot();
    });
    
});