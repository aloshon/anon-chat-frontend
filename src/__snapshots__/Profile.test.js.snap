// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile should render username and block list matches snapshot 1`] = `<DocumentFragment />`;
